pkg/steps/testing.go

```diff
@@ -5,21 +5,45 @@ package steps
 import (
 	"context"
 	"reflect"
+	"sync"
 	"testing"
 
-	fakeimageclientset "github.com/openshift/client-go/image/clientset/versioned/fake"
-	imagev1 "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
-	fakeimagev1 "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1/fake"
-
 	v1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/diff"
+
 	"k8s.io/client-go/kubernetes/fake"
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	fakecorev1 "k8s.io/client-go/kubernetes/typed/core/v1/fake"
 
+	fakeimageclientset "github.com/openshift/client-go/image/clientset/versioned/fake"
+	imagev1 "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
+	fakeimagev1 "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1/fake"
+
 	"github.com/openshift/ci-tools/pkg/api"
 )
 
+// DryLogger holds the information of all objects that have been created from a dry run.
+type DryLogger struct {
+	sync.RWMutex
+	objects []runtime.Object
+}
+
+// AddObject is adding an object to the list.
+func (dl *DryLogger) AddObject(o runtime.Object) {
+	dl.Lock()
+	defer dl.Unlock()
+	dl.objects = append(dl.objects, o)
+}
+
+// GetObjects returns the list of objects.
+func (dl *DryLogger) GetObjects() []runtime.Object {
+	dl.RLock()
+	defer dl.RUnlock()
```

Comment:
Is this actually safe though? We return just the header for writing but the underlying memory is the same...

Reply:
If we assign the objects in `x` with `GetObjects()` and it, the `DryLogger` won't change.
example: https://play.golang.org/p/S-heq951JOy

Reply:
I think Petr's question was about the opposite case, `AddObjects` after a slice has been obtained with `GetObjects`.

As I understand it, slices are a `(pointer, length)` pair, so even if `objects` were modified (and even if it had to be reallocated to grow), the slice would still be (outdated, but) valid. Garbage collection should ensure the memory remains valid as long as the slice also is.

This does not guarantee that the objects in the slice cannot be modified, but I
don't think that is a problem here.